File: _includes/code/object-oriented-programming/virtual-dispatch/virtual.cc
#include <stdio.h>

struct Animal {
  virtual const char *name() const { return "Animal"; }
  virtual void move() const { printf("walk %s\n", name()); }
  virtual ~Animal() {}
};

struct Bird : public Animal {
  virtual void move() const override { printf("fly %s\n", name()); }
  virtual const char *name() const override { return "Bird"; }
};

struct Elephant : public Animal {
  virtual const char *name() const override { return "Elephant"; }
};

int main(void) {
  const Animal *animals[] = {new Animal(), new Bird(), new Elephant()};

  for (const auto &animal : animals) {
    animal->move();
  }
}

File: _includes/code/object-oriented-programming/virtual-dispatch/virtual.py
class Animal:
    def name(self):
        return "Animal"

    def move(self):
        print(f"walk {self.name()}")


class Bird(Animal):
    def name(self):
        return "Bird"

    def move(self):
        print(f"fly {self.name()}")


class Elephant(Animal):
    def name(self):
        return "Elephant"


def main():
    animals = [Animal(), Bird(), Elephant()]
    for animal in animals:
        animal.move()


if __name__ == "__main__":
    main()
